vllm/entrypoints/openai/protocol.py

```diff
@@ -300,21 +358,63 @@ def check_guided_decoding_count(cls, data):
-                "tool_choice"] != "none":
+        if guide_count > 1 and data.get('tool_choice',
+                                        'none') not in ("none", "auto"):
             raise ValueError(
                 "You can only either use guided decoding or tools, not both.")
         return data
 
     @model_validator(mode="before")
     @classmethod
-    def check_tool_choice(cls, data):
-        if "tool_choice" in data and data["tool_choice"] != "none":
-            if not isinstance(data["tool_choice"], dict):
-                raise ValueError("Currently only named tools are supported.")
+    def check_tool_usage(cls, data):
+
+        # if "tool_choice" is not specified but tools are provided,
+        # default to "auto" tool_choice
+        if "tool_choice" not in data and "tools" in data:
+            data["tool_choice"] = "auto"
+
+        # if "tool_choice" is specified -- validation
+        if "tool_choice" in data:
+
+            # ensure that if "tool choice" is specified, tools are present
             if "tools" not in data or data["tools"] is None:
                 raise ValueError(
                     "When using `tool_choice`, `tools` must be set.")
+
+            # make sure that tool choice is either a named tool
+            # OR that it's set to "auto"
+            if data["tool_choice"] != "auto" and not isinstance(
+                    data["tool_choice"], dict):
+                raise ValueError(
+                    "`tool_choice` must either be a named tool or \"auto\". "
+                    "`tool_choice=\"none\" is not supported.")
+
+            # ensure that if "tool_choice" is specified as an object,
+            # it matches a valid tool
+            if isinstance(data["tool_choice"], dict):
+                valid_tool = False
+                specified_function = data["tool_choice"]["function"]
+                if not specified_function:
+                    return ValueError(
+                        'Incorrectly formatted `tool_choice`. Should be like '
+                        '`{"type": "function",'
+                        ' "function": {"name": "my_function"}}`')
+                specified_function_name = specified_function["name"]
+                if not specified_function_name:
+                    return ValueError(
+                        'Incorrectly formatted `tool_choice`. Should be like '
+                        '`{"type": "function", '
+                        '"function": {"name": "my_function"}}`')
+                for tool in data['tools']:
+                    if tool["function"]["name"] == specified_function_name:
+                        valid_tool = True
+                        break
+                if not valid_tool:
+                    return ValueError(
+                        "The tool specified in `tool_choice` does not match any"
+                        " of the specified `tools`")
+
+        # TODO validate tools
```

Comment:
what else should be done here?

Reply:
I thought about this and making sure that the tools passed into the server match the OpenAI format but since all that happens is that these get passed into the chat template, it seemed to me like we should allow the user to pass in a differently-structured tool if that's what their model & chat template use, just like we allow non-standard chat roles for models & chat templates that use it. 

Let me know if you feel differently and I can validate them! Otherwise, I'll remove the comment for now.